src/qibo/tomography_RGD/qibo_states.py

```diff
@@ -0,0 +1,257 @@
+
+            probe_circuit.add(gates.M(*qubits))
+
+            measurement_circuit_name = self.make_measurement_circuit_name(
+                self.circuit_name, label
+            )
+            measurement_circuit = self.circuit + probe_circuit
+            self.measurement_circuit_names.append(measurement_circuit_name)
+            self.measurement_circuits.append(measurement_circuit)
+
+    @staticmethod
+    def make_measurement_circuit_name(circuit_name, label):
+        """
+        Measurement circuit naming convention
+        """
+        name = "{}-{}".format(circuit_name, label)
+        return name
+
+    @staticmethod
+    def sort_sample(sample) -> dict:
+        """
+        Sort raw measurements into count dictionary
+        """
+        sorted_sample = {}
+        for shot in sample:
+            s = "".join(map(str, shot))
+            if s in sorted_sample:
+                sorted_sample[s] += 1
+            else:
+                sorted_sample[s] = 1
+        return sorted_sample
+
+    def execute_measurement_circuits(
+        self,
+        labels,
+        backend="numpy",
+        num_shots=100,
+        # num_shots = 10000,
+        label_format="little_endian",
+    ):
+        """
+        Executes measurement circuits
+        - labels: string of Pauli matrices (e.g. XYZXX)
+        - backend: 'numpy', 'qibojit', 'pytorch', 'tensorflow' prvided by qibo
+        - num_shots: number of shots measurement is taken to get empirical frequency through counts
+        """
+        if self.measurement_circuit_names == []:
+            self.create_measurement_circuits(labels, label_format)
+
+        circuit_names = self.measurement_circuit_names
+
+        qibo.set_backend(backend)
+
+        data_dict_list = []
+        for i, label in enumerate(labels):
+            result = self.measurement_circuits[i].execute(nshots=num_shots)
+            result.samples(binary=True, registers=False)
+            count_dict = self.sort_sample(result.to_dict()["samples"])
+
+            measurement_circuit_name = self.make_measurement_circuit_name(
+                self.circuit_name, label
+            )
+            data_dict = {
+                "measurement_circuit_name": measurement_circuit_name,
+                "circuit_name": self.circuit_name,
+                "label": label,
+                "count_dict": count_dict,
+                "backend": "qibo: " + backend,
+                "num_shots": num_shots,
+            }
+            data_dict_list.append(data_dict)
+        return data_dict_list
+
+
+class GHZState(State):
+    """
+    Constructor for GHZState class
+    """
+
+    def __init__(self, n):
+        State.__init__(self, n)
+        self.circuit_name = "GHZ"
+
+    def create_circuit(self):
+        circuit = qibo.Circuit(self.n)
+
+        circuit.add(gates.H(0))
+        for i in range(1, self.n):
+            circuit.add(gates.CNOT(0, i))
+
+        self.circuit = circuit

```

Comment:
This fits as a function in `qibo.models.encodings` receiving `nqubits` as input and returning the corresponding circuit as output.

Reply:
A new branch starting from `master`, completely unrelated to this one. You can name whatever, but probably `ghz` is simpler and better. Then in that branch, you only implement the GHZ circuit as a function in `qibo.models.encodings` and the corresponding tests. Nothing else. After that is merged into `master`, we change the code here accordingly so it can communicate with the newly existing function.

Reply:
Thank you for the instruction. I have created a new branch called ghz and push it with sync fork. I have tested some functions inside the directory src/qibo/test. However, I do not understand well the qibo.models.encodings you mentioned. There are some encoders in the qibo.models.encodings but probably we don't need them at the moment. I am not sure if we can add some user-defined properties in the circuit. The way I did it now is to create a new class. Probably some properties do not fit into qibo's standards. I am curious if creating some list as a recorded processed result is allowed. Or do you prefer on-the-fly calculation during processing? Thanks.

Reply:
I think what @renatomello is that the GHZ state should be added as a function to `qibo.models.encodings`. It will probably look something like this:

``` python
def GHZ_circuit(nqubits):
    c = QiboCircuit(nqubits)
    c.add(gates.H(0))
    for _i in range(nqubits-1):
        c.add(gates.CNOT(_i, _i+1))
    return c
```

I don't think it will take so much time to add it in, I can do it after dinner.
Anyway, would `qibo.models.utils` be a more suitable location than  `qibo.models.encodings`?